### PR TITLE
[NVPTX] support switch statement with brx.idx

### PR DESCRIPTION
**Pull request by AlexMaclean**
Add custom lowering for `BR_JT` DAG nodes to the `brx.idx` PTX instruction ([PTX ISA 9.7.13.4. Control Flow Instructions: brx.idx](https://docs.nvidia.com/cuda/parallel-thread-execution/#control-flow-instructions-brx-idx)). Depending on the heuristics in DAG selection, `switch` statements may now be lowered using `brx.idx`

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td [hunk 1]
```diff
@@ -3880,6 +3880,44 @@ def DYNAMIC_STACKALLOC64 :
...
+let isTerminator = 1, isBranch = 1, isIndirectBranch = 1 in {
+
+  def BRX_START :
+    NVPTXInst<(outs), (ins i32imm:$id),
+              "$$L_brx_$id: .branchtargets",
+              [(brx_start (i32 imm:$id))]>;
+
+  def BRX_ITEM :
+    NVPTXInst<(outs), (ins brtarget:$target),
+              "$target,",
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3911]: Nit: We may want to indent the labels in the list. Right now they seem to end up aligned on the instruction boundary, while they are actually arguments of the `.branchtargets` above. Does not impact functionality, but it looks somewhat odd.

**Review by Artem-B — APPROVED** [review]
LGTM.

Does brx.idx buy us a better code on SASS level? or is this mostly cosmetic sugar on PTX level?

**Comment by AlexMaclean** [comment]
> Does brx.idx buy us a better code on SASS level? or is this mostly cosmetic sugar on PTX level?

This does lead to improvements on the SASS level in some cases, though this is of course very dependent on `ptxas` optimizations.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 1 above)
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Fixed

**Review by Artem-B — APPROVED** [review]

**Pull request merged**